gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ReplicationUtils.java

```diff
@@ -0,0 +1,157 @@
+      return new ReplicationMetaData(Optional.<Map<String, String>> absent());
+    }
+
+    Config metaDataConfig = config.getConfig(METADATA);
+    Map<String, String> metaDataValues = new HashMap<>();
+    Set<Map.Entry<String, ConfigValue>> meataDataEntry = metaDataConfig.entrySet();
+    for (Map.Entry<String, ConfigValue> entry : meataDataEntry) {
+      metaDataValues.put(entry.getKey(), metaDataConfig.getString(entry.getKey()));
+    }
+
+    ReplicationMetaData metaData = new ReplicationMetaData(Optional.of(metaDataValues));
+    return metaData;
+  }
+
+  public static SourceEndPoint buildSource(Config config) {
+    Preconditions.checkArgument(config.hasPath(REPLICATION_SOURCE),
+        "missing required config entery " + REPLICATION_SOURCE);
+
+    Config sourceConfig = config.getConfig(REPLICATION_SOURCE);
+
+    return new SourceEndPoint(ReplicationUtils.buildReplicationLocation(sourceConfig),
+        ReplicationUtils.buildDataset(sourceConfig));
+  }
+
+  public static List<ReplicaEndPoint> buildReplicas(Config config) {
+    Preconditions.checkArgument(config.hasPath(REPLICATION_REPLICAS),
+        "missing required config entery " + REPLICATION_REPLICAS);
+
+    Config replicaConfig = config.getConfig(REPLICATION_REPLICAS);
+    Preconditions.checkArgument(replicaConfig.hasPath(REPLICATOIN_REPLICAS_LIST),
+        "missing required config entery " + REPLICATOIN_REPLICAS_LIST);
+    List<String> replicas = replicaConfig.getStringList(REPLICATOIN_REPLICAS_LIST);
+
+    List<ReplicaEndPoint> result = new ArrayList<ReplicaEndPoint>();
+    for (String replicaName : replicas) {
+      Config subConfig = replicaConfig.getConfig(replicaName);
+      result.add(new ReplicaEndPoint(replicaName, ReplicationUtils.buildReplicationLocation(subConfig),
+          ReplicationUtils.buildDataset(subConfig)));
+    }
+
+    return result;
+  }
+
+  public static DataFlowTopology buildDataFlowTopology(Config config, SourceEndPoint source,
+      List<ReplicaEndPoint> replicas) {
+    Preconditions.checkArgument(config.hasPath(DATAFLOWTOPOLOGY), "missing required config entery " + DATAFLOWTOPOLOGY);
+    Config dataflowConfig = config.getConfig(DATAFLOWTOPOLOGY);
+
+    // NOT specified by literal routes, need to pick it
+    if (!dataflowConfig.hasPath(DataFlowTopology.ROUTES)) {
+      // DEFAULT_ALL_ROUTES_KEY specified in top level config
+      Preconditions.checkArgument(config.hasPath(DEFAULT_ALL_ROUTES_KEY),
+          "missing required config entery " + DEFAULT_ALL_ROUTES_KEY);
+
+      Config allRoutes = config.getConfig(DEFAULT_ALL_ROUTES_KEY);
+      String routePickerClassName = dataflowConfig.hasPath(ROUTES_PICKER_CLASS_KEY)
+          ? dataflowConfig.getString(ROUTES_PICKER_CLASS_KEY) : DEFAULT_ROUTES_PICKER_CLASS_KEY;
+
+      List<Object> args = Lists.newArrayList(allRoutes, source);
+
+      try {
+        Class<?> routePickerClass = Class.forName(routePickerClassName);
+
+        DataFlowRoutesPicker picker =
```

Comment:
A couple of meetings ago, we discussed in that there should not be invocation of non-default constructors through reflection.